source/initial_temperature/patch_on_S40RTS.cc

```diff
@@ -0,0 +1,206 @@
+{
+  namespace InitialTemperature
+  {
+    template <int dim>
+    PatchOnS40RTS<dim>::PatchOnS40RTS ()
+    {}
+
+
+    template <int dim>
+    void
+    PatchOnS40RTS<dim>::initialize ()
+    {
+      Utilities::AsciiDataInitial<dim>::initialize(1);
+      if (s40rts.vs_to_density_method == s40rts.file)
+        {
+          s40rts.profile.initialize(this->get_mpi_communicator());
+          s40rts.vs_to_density_index = s40rts.profile.get_column_index_from_name("vs_to_density");
+        }
+    }
+
+
+    template <int dim>
+    double
+    PatchOnS40RTS<dim>::
+    ascii_grid_vs (const Point<dim> &position) const
+    {
+      const double vs_perturbation = Utilities::AsciiDataInitial<dim>::get_data_component(position,0);
+      return vs_perturbation;
+    }
+
+
+    template <int dim>
+    double
+    PatchOnS40RTS<dim>::
+    initial_temperature (const Point<dim> &position) const
+    {
+      const double depth = this->get_geometry_model().depth(position);
+
+      double vs_perturbation;
+      if (depth <= max_grid_depth - smoothing_length_scale)
+        {
+          vs_perturbation = ascii_grid_vs(position);
+        }
+      //add smoothing between the two models
+      else if (depth > max_grid_depth - smoothing_length_scale && depth < max_grid_depth)
+        {
+          const double scale_factor = (depth-(max_grid_depth-smoothing_length_scale))/smoothing_length_scale;
+          vs_perturbation = s40rts.get_Vs(position)*(scale_factor) + ascii_grid_vs(position)*(1.0-scale_factor);
+        }
+      else
+        {
+          vs_perturbation = s40rts.get_Vs(position);
+        }
+
+      // use either the user-input reference temperature as background temperature
+      // (incompressible model) or the adiabatic temperature profile (compressible model)
+      const double background_temperature = this->get_material_model().is_compressible() ?
+                                            this->get_adiabatic_conditions().temperature(position) :
+                                            s40rts.reference_temperature;
+
+      // get the Vs to density conversion
+      double vs_to_density = 0.0;
+      if (s40rts.vs_to_density_method == s40rts.file)
+        vs_to_density = s40rts.profile.get_data_component(Point<1>(depth), s40rts.vs_to_density_index);
+      else if (s40rts.vs_to_density_method == s40rts.constant)
+        vs_to_density = s40rts.vs_to_density_constant;
+      else
+        // we shouldn't get here but instead should already have been
+        // kicked out when declaring the parameter, as the pattern won't match ("file|constant")
+        Assert(false, ExcMessage("Unknown method for vs to density scaling."));
+
+      // scale the perturbation in seismic velocity into a density perturbation
+      // vs_to_density is read in from input file
+      const double density_perturbation = vs_to_density * vs_perturbation;
+
+      double temperature_perturbation;
+      if (depth > s40rts.no_perturbation_depth)
```

Comment:
Unfortunately I can't find my comment about this `if` statement any more. It seems unclear to me that the `s40rts.no_perturbation_depth` should have effect on the perturbation of the patch. I think it should only work on the s40rts perturbation (e.g. in line 79), or be removed completely (as the upper part of s40rts is replaced by the patch anyway), or there should be a separate no_perturbation_depth for the upper mantle model.

Reply:
Sorry this is causing confusion. 
I've tested it and the no_perturbation_depth does definitely work on the ascii grid too, we are just saying that any temperature perturbation above this depth (S40RTS or the replaced ascii grid) is zeroed out. 
This is exactly what we want if we are trying to remove the contribution from the density contrasts in the lithosphere (i.e. to zero out the top 100 or 200 km). Although this does zero out some of the new data we have put in with the ascii grid, we still have a lot of it left as this upper mantle tomography goes to a depth of 700 km.
@jaustermann am I explaining this correctly? Do you agree?

I guess what may make this slightly clearer is if we have a no_perturbation_depth parameter for this material model as a whole and declare it in patch_on_S40RTS rather than parsing it from S40RTS_perturbation? Would this help?

Reply:
I agree @SophieCoulson, I think it is currently doing the right thing but I also agree that this could be confusing. Maybe making no_perturbation_depth a parameter within patch_on_S40RTS rather than S40RTS_perturbation (as you suggest Sophie) is the way to go.